pkg/flowaggregator/flowaggregator.go

```diff
@@ -15,25 +15,257 @@
+		"egressNetworkPolicyName",
+		"egressNetworkPolicyNamespace",
+		"destinationClusterIPv4",
+	}
+	aggregatorElements = []string{
+		"originalExporterIPv4Address",
+		"originalObservationDomainId",
+	}
+	aggregationFields = []string{
+		"destinationPodName",
+		"destinationPodNamespace",
+		"destinationNodeName",
+		"destinationClusterIPv4",
+		"destinationServicePortName",
+		"ingressNetworkPolicyName",
+		"ingressNetworkPolicyNamespace",
+	}
+	aggregationWorkerNum = 2
+)
+
+type AggregatorTransportProtocol string
+
+const (
+	AggregatorTransportProtocolTCP AggregatorTransportProtocol = "TCP"
+	AggregatorTransportProtocolUDP AggregatorTransportProtocol = "UDP"
 )
 
 type flowAggregator struct {
-	flowCollectorAddr net.Addr
-	exportInterval    time.Duration
+	externalFlowCollectorAddr   net.Addr
+	aggregatorTransportProtocol AggregatorTransportProtocol
+	collectingProcess           ipfix.IPFIXCollectingProcess
+	aggregationProcess          ipfix.IPFIXAggregationProcess
+	exportInterval              time.Duration
+	exportingProcess            ipfix.IPFIXExportingProcess
+	templateID                  uint16
+	registry                    ipfix.IPFIXRegistry
 }
 
-func InitFlowAggregator(flowCollectorAddr net.Addr, exportInterval time.Duration) *flowAggregator {
+func NewFlowAggregator(externalFlowCollectorAddr net.Addr, exportInterval time.Duration, aggregatorTransportProtocol AggregatorTransportProtocol) *flowAggregator {
+	registry := ipfix.NewIPFIXRegistry()
 	registry.LoadRegistry()
-	return &flowAggregator{
-		flowCollectorAddr,
+	fa := &flowAggregator{
+		externalFlowCollectorAddr,
+		aggregatorTransportProtocol,
+		nil,
+		nil,
 		exportInterval,
+		nil,
+		0,
+		registry,
+	}
+	return fa
+}
+
+func genObservationID() (uint32, error) {
+	name := env.GetPodName()
+	if name == "" {
+		return 0, fmt.Errorf("env.GetPodName returns empty pod name")
+	}
+	h := fnv.New32()
+	h.Write([]byte(name))
+	return h.Sum32(), nil
+}
+
+func (fa *flowAggregator) InitCollectingProcess() error {
+	var collectAddr net.Addr
+	var err error
+	if fa.aggregatorTransportProtocol == AggregatorTransportProtocolTCP {
+		collectAddr, _ = net.ResolveTCPAddr("tcp", "0.0.0.0:4739")
+		fa.collectingProcess, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 65535, 0)
+	} else {
+		collectAddr, _ = net.ResolveUDPAddr("udp", "0.0.0.0:4739")
+		fa.collectingProcess, err = ipfix.NewIPFIXCollectingProcess(collectAddr, 1024, 0)
+	}
+	return err
+}
+
+func (fa *flowAggregator) InitAggregationProcess() error {
+	var err error
+	fa.aggregationProcess, err = ipfix.NewIPFIXAggregationProcess(fa.collectingProcess.GetMsgChan(), aggregationWorkerNum, aggregationFields)
+	return err
+}
+
+func (fa *flowAggregator) initExportingProcess() error {
+	obsID, err := genObservationID()
+	if err != nil {
+		return fmt.Errorf("cannot generate observation ID for flow aggregator: %v", err)
+	}
+	var ep ipfix.IPFIXExportingProcess
+	if fa.externalFlowCollectorAddr.Network() == "tcp" {
```

Comment:
use `fa.aggregatorTransportProtocol` here instead?

Reply:
We are checking the transport protocol of external flow collector here, so better not to use default protocol of aggregator collecting process to avoid misunderstanding.

Reply:
My bad, you're right